File: exercicios/aula13/ex03_ast_s_atribuida/ast_s.py
"""Exercício 03 — construtor de AST com SDD S-atribuída.

Gramática:

    E -> E + T | E - T | T
    T -> T * F | T / F | F
    F -> - F | ( E ) | id | num

Um único atributo, `node`, sintetizado em todos os não terminais. Cada ação
usa **só** atributos dos filhos (por isso a SDD é S-atribuída):

    E -> E1 op T   E.node = Binary(op, E1.node, T.node)
    E -> T         E.node = T.node
    T -> T1 op F   T.node = Binary(op, T1.node, F.node)
    T -> F         T.node = F.node
    F -> - F1      F.node = Unary('-', F1.node)
    F -> ( E )     F.node = E.node          (parênteses não geram nó)
    F -> id        F.node = Identifier(id.lexema)
    F -> num       F.node = Number(num.lexema)

Os laços de E e T reduzem da esquerda para a direita, então em `a - 4 + c` o
Binary('-') vira o filho esquerdo do Binary('+'): (+ (- a 4) c).

Duas travessias sobre a árvore pronta: impressão prefixa totalmente
parentizada e avaliação (inteiros de 64 bits, como no C). Uma expressão com
identificador não é avaliável — isso não é erro, só não há valor para mostrar.

Uso:  python ast_s.py [--trace] <arquivo | ->
"""

import sys
from dataclasses import dataclass
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1] / "comum"))
import fluxo as F  # noqa: E402

MIN, MAX = -(2**63), 2**63 - 1


# -- nós da AST (imutáveis) ------------------------------------------------


@dataclass(frozen=True)
class Number:
    lexema: str
    linha: int
    coluna: int


@dataclass(frozen=True)
class Identifier:
    nome: str
    linha: int
    coluna: int


@dataclass(frozen=True)
class Unary:
    op: str
    operando: "No"
    linha: int
    coluna: int


@dataclass(frozen=True)
class Binary:
    op: str
    esq: "No"
    dir: "No"
    linha: int
    coluna: int


No = Number | Identifier | Unary | Binary


# -- parser com as ações semânticas ----------------------------------------


class Construtor:
    def __init__(self, fl: F.Fluxo, trace: bool):
        self.fl = fl
        self.trace = trace

    def _regra(self, producao: str, acao: str):
        if self.trace:
            print(f"{producao:<14}{acao}")

    def parse_E(self) -> No:
        node = self.parse_T()
        self._regra("E -> T", "E.node = T.node")
        while self.fl.verifica("PLUS", "MINUS"):
            op = self.fl.proximo()
            direita = self.parse_T()
            node = Binary(op.lexeme, node, direita, op.line, op.column)
            self._regra(f"E -> E {op.lexeme} T", f"E.node = Binary('{op.lexeme}', E1.node, T.node)")
        return node

    def parse_T(self) -> No:
        node = self.parse_F()
        self._regra("T -> F", "T.node = F.node")
        while self.fl.verifica("STAR", "SLASH"):
            op = self.fl.proximo()
            direita = self.parse_F()
            node = Binary(op.lexeme, node, direita, op.line, op.column)
            self._regra(f"T -> T {op.lexeme} F", f"T.node = Binary('{op.lexeme}', T1.node, F.node)")
        return node

    def parse_F(self) -> No:
        tok = self.fl.atual()
        if self.fl.aceita("MINUS"):
            operando = self.parse_F()
            self._regra("F -> - F", "F.node = Unary('-', F1.node)")
            return Unary("-", operando, tok.line, tok.column)
        if self.fl.aceita("LPAREN"):
            node = self.parse_E()
            self.fl.exige("RPAREN", "')'")
            self._regra("F -> ( E )", "F.node = E.node")
            return node
        if self.fl.aceita("IDENT"):
            self._regra("F -> id", f"F.node = Identifier({tok.lexeme})")
            return Identifier(tok.lexeme, tok.line, tok.column)
        if self.fl.aceita("INT"):
            self._regra("F -> num", f"F.node = Number({tok.lexeme})")
            return Number(tok.lexeme, tok.line, tok.column)
        raise F.ErroSintaxe(tok, "número, identificador, '-' ou '('")


# -- travessias ------------------------------------------------------------


def prefixa(no: No) -> str:
    match no:
        case Number(lexema=lex):
            return lex
        case Identifier(nome=nome):
            return nome
        case Unary(op=op, operando=o):
            return f"({op} {prefixa(o)})"
        case Binary(op=op, esq=e, dir=d):
            return f"({op} {prefixa(e)} {prefixa(d)})"


class NaoAvaliavel(Exception):
    """A expressão tem identificador: não há valor numérico (não é erro)."""


class ErroAvaliacao(Exception):
    """Erro semântico durante a avaliação (divisão por zero, estouro)."""


def avalia(no: No) -> int:
    match no:
        case Number(lexema=lex, linha=l, coluna=c):
            v = int(lex)
            if v > MAX:
                raise ErroAvaliacao(F.erro_semantico(l, c, f"número {lex} fora do intervalo de 64 bits"))
            return v
        case Identifier(nome=nome, linha=l, coluna=c):
            raise NaoAvaliavel(f"identificador '{nome}' na linha {l}, coluna {c}")
        case Unary(operando=o, linha=l, coluna=c):
            v = -avalia(o)
            if v > MAX:
                raise ErroAvaliacao(F.erro_semantico(l, c, "estouro de inteiro de 64 bits"))
            return v
        case Binary(op=op, esq=e, dir=d, linha=l, coluna=c):
            a, b = avalia(e), avalia(d)
            if op == "/":
                if b == 0:
                    raise ErroAvaliacao(F.erro_semantico(l, c, f"divisão por zero ({a} / {b})"))
                q = abs(a) // abs(b)
                r = -q if (a < 0) != (b < 0) else q
            else:
                r = {"+": a + b, "-": a - b, "*": a * b}[op]
            if not MIN <= r <= MAX:
                raise ErroAvaliacao(F.erro_semantico(l, c, f"estouro de inteiro de 64 bits ({a} {op} {b})"))
            return r


def main(argv: list[str]) -> int:
    F.configurar_saida()
    trace = "--trace" in argv[1:]
    args = [a for a in argv[1:] if a != "--trace"]
    if len(args) != 1:
        print("uso: python ast_s.py [--trace] <arquivo | ->", file=sys.stderr)
        return F.ERRO_USO

    fl, codigo = F.abrir(args[0])
    if fl is None:
        return codigo
    try:
        arvore = Construtor(fl, trace).parse_E()
        if not fl.verifica("EOF"):
            raise F.ErroSintaxe(fl.atual(), "operador ou fim da entrada")
    except F.ErroSintaxe as erro:
        print(erro, file=sys.stderr)
        return F.ERRO_SINTAXE

    print(f"prefixa: {prefixa(arvore)}")
    try:
        print(f"valor: {avalia(arvore)}")
    except NaoAvaliavel as motivo:
        print(f"valor: não avaliável ({motivo})")
    except ErroAvaliacao as erro:
        print(erro, file=sys.stderr)
        return F.ERRO_SEMANTICO
    return F.OK


if __name__ == "__main__":
    sys.exit(main(sys.argv))

File: exercicios/aula13/ex03_ast_s_atribuida/ast_s.c
/* Exercício 03 — construtor de AST com SDD S-atribuída.
 *
 * Mesma gramática, mesmas ações e mesma saída de ast_s.py (ver a docstring de
 * lá). Representação em C:
 *
 * - `No` é uma struct discriminada por `NoTipo` (enum) com uma union para os
 *   campos de cada tipo;
 * - **posse**: cada nó é dono dos seus filhos e do seu texto. `no_free` libera
 *   a subárvore inteira;
 * - **falha de alocação**: os construtores recebem os filhos já construídos e,
 *   se o malloc falhar, liberam esses filhos antes de devolver NULL. Assim quem
 *   chama nunca precisa limpar nada num caminho de erro: basta propagar NULL.
 *   O parser distingue os dois motivos de NULL (sintaxe ou memória) pelo campo
 *   `status`.
 *
 * Uso:  ./ast_s [--trace] <arquivo | ->
 */

#include <errno.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "../comum/fluxo.h"

typedef enum { NO_NUMBER, NO_IDENTIFIER, NO_UNARY, NO_BINARY } NoTipo;

typedef struct No {
    NoTipo tipo;
    int linha, coluna;
    union {
        char *texto; /* NUMBER: o lexema; IDENTIFIER: o nome */
        struct {
            char op;
            struct No *operando;
        } un;
        struct {
            char op;
            struct No *esq, *dir;
        } bin;
    } u;
} No;

static void no_free(No *n) {
    if (!n) return;
    switch (n->tipo) {
    case NO_NUMBER:
    case NO_IDENTIFIER: free(n->u.texto); break;
    case NO_UNARY: no_free(n->u.un.operando); break;
    case NO_BINARY: no_free(n->u.bin.esq), no_free(n->u.bin.dir); break;
    }
    free(n);
}

/* -- construtores: devolvem NULL se faltar memória, sem vazar nada -------- */

static No *novo(NoTipo tipo, const Token *t) {
    No *n = malloc(sizeof(No));
    if (n) {
        n->tipo = tipo;
        n->linha = t->line;
        n->coluna = t->column;
    }
    return n;
}

static No *folha(NoTipo tipo, const Token *t) {
    No *n = novo(tipo, t);
    char *txt = malloc(strlen(t->lexeme) + 1);
    if (!n || !txt) {
        free(n), free(txt);
        return NULL;
    }
    strcpy(txt, t->lexeme);
    n->u.texto = txt;
    return n;
}

static No *mk_unary(const Token *op, No *operando) {
    No *n = operando ? novo(NO_UNARY, op) : NULL;
    if (!n) {
        no_free(operando); /* o construtor é dono do filho, mesmo falhando */
        return NULL;
    }
    n->u.un.op = op->lexeme[0];
    n->u.un.operando = operando;
    return n;
}

static No *mk_binary(const Token *op, No *esq, No *dir) {
    No *n = (esq && dir) ? novo(NO_BINARY, op) : NULL;
    if (!n) {
        no_free(esq), no_free(dir);
        return NULL;
    }
    n->u.bin.op = op->lexeme[0];
    n->u.bin.esq = esq;
    n->u.bin.dir = dir;
    return n;
}

/* -- parser com as ações semânticas ---------------------------------------- */

typedef enum { ST_OK, ST_SINTAXE, ST_MEMORIA } Status;

typedef struct {
    Fluxo *fl;
    int trace;
    Status status;
} Construtor;

static void regra(const Construtor *c, const char *prod, const char *acao) {
    if (c->trace) printf("%-14s%s\n", prod, acao);
}

/* Após um construtor devolver NULL sem erro de sintaxe, a causa é memória. */
static No *confere(Construtor *c, No *n) {
    if (!n && c->status == ST_OK) c->status = ST_MEMORIA;
    return n;
}

static No *parse_E(Construtor *c);

static No *parse_F(Construtor *c) {
    const Token *t = fluxo_atual(c->fl);
    if (fluxo_aceita(c->fl, TOK_MINUS)) {
        No *o = parse_F(c);
        if (!o) return NULL;
        regra(c, "F -> - F", "F.node = Unary('-', F1.node)");
        return confere(c, mk_unary(t, o));
    }
    if (fluxo_aceita(c->fl, TOK_LPAREN)) {
        No *n = parse_E(c);
        if (!n) return NULL;
        if (!fluxo_verifica(c->fl, TOK_RPAREN)) {
            erro_sintaxe(fluxo_atual(c->fl), "')'");
            c->status = ST_SINTAXE;
            no_free(n);
            return NULL;
        }
        fluxo_proximo(c->fl);
        regra(c, "F -> ( E )", "F.node = E.node");
        return n;
    }
    if (t->type == TOK_IDENT || t->type == TOK_INT) {
        fluxo_proximo(c->fl);
        char *acao = t->type == TOK_IDENT ? formata("F.node = Identifier(%s)", t->lexeme)
                                          : formata("F.node = Number(%s)", t->lexeme);
        regra(c, t->type == TOK_IDENT ? "F -> id" : "F -> num", acao);
        free(acao);
        return confere(c, folha(t->type == TOK_IDENT ? NO_IDENTIFIER : NO_NUMBER, t));
    }
    erro_sintaxe(t, "número, identificador, '-' ou '('");
    c->status = ST_SINTAXE;
    return NULL;
}

static No *parse_T(Construtor *c) {
    No *n = parse_F(c);
    if (!n) return NULL;
    regra(c, "T -> F", "T.node = F.node");
    while (fluxo_verifica(c->fl, TOK_STAR) || fluxo_verifica(c->fl, TOK_SLASH)) {
        const Token *op = fluxo_proximo(c->fl);
        No *dir = parse_F(c);
        if (!dir) {
            no_free(n);
            return NULL;
        }
        n = confere(c, mk_binary(op, n, dir));
        if (!n) return NULL;
        char *prod = formata("T -> T %s F", op->lexeme);
        char *acao = formata("T.node = Binary('%s', T1.node, F.node)", op->lexeme);
        regra(c, prod, acao);
        free(prod), free(acao);
    }
    return n;
}

static No *parse_E(Construtor *c) {
    No *n = parse_T(c);
    if (!n) return NULL;
    regra(c, "E -> T", "E.node = T.node");
    while (fluxo_verifica(c->fl, TOK_PLUS) || fluxo_verifica(c->fl, TOK_MINUS)) {
        const Token *op = fluxo_proximo(c->fl);
        No *dir = parse_T(c);
        if (!dir) {
            no_free(n);
            return NULL;
        }
        n = confere(c, mk_binary(op, n, dir));
        if (!n) return NULL;
        char *prod = formata("E -> E %s T", op->lexeme);
        char *acao = formata("E.node = Binary('%s', E1.node, T.node)", op->lexeme);
        regra(c, prod, acao);
        free(prod), free(acao);
    }
    return n;
}

/* -- travessias ------------------------------------------------------------- */

static void prefixa(const No *n) {
    switch (n->tipo) {
    case NO_NUMBER:
    case NO_IDENTIFIER: fputs(n->u.texto, stdout); break;
    case NO_UNARY:
        printf("(%c ", n->u.un.op);
        prefixa(n->u.un.operando);
        putchar(')');
        break;
    case NO_BINARY:
        printf("(%c ", n->u.bin.op);
        prefixa(n->u.bin.esq);
        putchar(' ');
        prefixa(n->u.bin.dir);
        putchar(')');
        break;
    }
}

typedef enum { AV_OK, AV_NAO_AVALIAVEL, AV_ERRO } AvStatus;

/* Em AV_NAO_AVALIAVEL, *motivo aponta o identificador; em AV_ERRO a mensagem
 * já foi impressa em stderr. */
static AvStatus avalia(const No *n, long long *out, const No **motivo) {
    long long a, b;
    AvStatus s;
    switch (n->tipo) {
    case NO_NUMBER:
        errno = 0;
        *out = strtoll(n->u.texto, NULL, 10);
        if (errno == ERANGE) {
            fprintf(stderr, "Erro semântico na linha %d, coluna %d: número %s fora do intervalo de 64 bits.\n",
                    n->linha, n->coluna, n->u.texto);
            return AV_ERRO;
        }
        return AV_OK;
    case NO_IDENTIFIER: *motivo = n; return AV_NAO_AVALIAVEL;
    case NO_UNARY:
        if ((s = avalia(n->u.un.operando, &a, motivo)) != AV_OK) return s;
        if (a == LLONG_MIN) {
            fprintf(stderr, "Erro semântico na linha %d, coluna %d: estouro de inteiro de 64 bits.\n", n->linha,
                    n->coluna);
            return AV_ERRO;
        }
        *out = -a;
        return AV_OK;
    case NO_BINARY: break;
    }
    if ((s = avalia(n->u.bin.esq, &a, motivo)) != AV_OK) return s;
    if ((s = avalia(n->u.bin.dir, &b, motivo)) != AV_OK) return s;
    char op = n->u.bin.op;
    int ok = 1;
    if (op == '/' && b == 0) {
        fprintf(stderr, "Erro semântico na linha %d, coluna %d: divisão por zero (%lld / %lld).\n", n->linha,
                n->coluna, a, b);
        return AV_ERRO;
    }
    switch (op) {
    case '+': ok = !((b > 0 && a > LLONG_MAX - b) || (b < 0 && a < LLONG_MIN - b)); if (ok) *out = a + b; break;
    case '-': ok = !((b < 0 && a > LLONG_MAX + b) || (b > 0 && a < LLONG_MIN + b)); if (ok) *out = a - b; break;
    case '*':
        if (a > 0) ok = b > 0 ? a <= LLONG_MAX / b : b >= LLONG_MIN / a;
        else if (a < 0) ok = b > 0 ? a >= LLONG_MIN / b : b >= LLONG_MAX / a;
        if (ok) *out = a * b;
        break;
    default: ok = !(a == LLONG_MIN && b == -1); if (ok) *out = a / b; break;
    }
    if (!ok) {
        fprintf(stderr, "Erro semântico na linha %d, coluna %d: estouro de inteiro de 64 bits (%lld %c %lld).\n",
                n->linha, n->coluna, a, op, b);
        return AV_ERRO;
    }
    return AV_OK;
}

int main(int argc, char **argv) {
    int trace = 0, n_args = 0;
    const char *arquivo = NULL;
    for (int i = 1; i < argc; i++) {
        if (strcmp(argv[i], "--trace") == 0) trace = 1;
        else arquivo = argv[i], n_args++;
    }
    if (n_args != 1) {
        fprintf(stderr, "uso: %s [--trace] <arquivo | ->\n", argv[0]);
        return ERRO_USO;
    }
    Fluxo fl;
    int codigo = fluxo_abrir(&fl, arquivo);
    if (codigo != OK) return codigo;

    Construtor c = {&fl, trace, ST_OK};
    No *arvore = parse_E(&c);
    if (arvore && !fluxo_verifica(&fl, TOK_EOF)) {
        erro_sintaxe(fluxo_atual(&fl), "operador ou fim da entrada");
        no_free(arvore);
        arvore = NULL;
        c.status = ST_SINTAXE;
    }
    if (!arvore) {
        if (c.status == ST_MEMORIA) fprintf(stderr, "erro: memória insuficiente ao construir a AST\n");
        fluxo_fechar(&fl);
        return c.status == ST_MEMORIA ? ERRO_USO : ERRO_SINTAXE;
    }

    fputs("prefixa: ", stdout);
    prefixa(arvore);
    putchar('\n');
    fflush(stdout);
    long long v = 0;
    const No *motivo = NULL;
    AvStatus s = avalia(arvore, &v, &motivo);
    if (s == AV_OK) {
        printf("valor: %lld\n", v);
    } else if (s == AV_NAO_AVALIAVEL) {
        printf("valor: não avaliável (identificador '%s' na linha %d, coluna %d)\n", motivo->u.texto, motivo->linha,
               motivo->coluna);
    } else {
        codigo = ERRO_SEMANTICO;
    }
    no_free(arvore);
    fluxo_fechar(&fl);
    return codigo;
}
